multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/OperationsApiServiceImpl.java

```diff
@@ -214,16 +214,33 @@ private Operation getOperation(String operationId) {
             Operation operation = operationService.createQuery()
                                                   .processId(operationId)
                                                   .singleResult();
-            operation = operationsHelper.addState(operation);
+
+            operation = addOperationState(operation);
+
             operation = operationsHelper.addErrorType(operation);
             return operation;
         } catch (NoResultException e) {
             throw new NotFoundException(e, Messages.OPERATION_0_NOT_FOUND, operationId);
         }
     }
 
+    private Operation addOperationState(Operation operation) {
+        long startTime = System.nanoTime();
+        operation = operationsHelper.addState(operation);
+        long totalTimeInNano = System.nanoTime() - startTime;
+        LOGGER.debug("Operation state calculation took {} nanoseconds.", totalTimeInNano);
+        if (operation.getCurrentState() == null) {
+            LOGGER.warn("Cached operation is null for process \"{}\"!", operation.getProcessId());
+        }
+        if (operation.getCurrentState() != operation.getFinalState()) {
```

Comment:
The same as above: if the values differ, this will mean that there might be a gap in the logic, which should be fixed. During the execution of the following takt, we should be aware of such logs. If there are such logs, it is bad and we should investigate the reason and fix or in worst scenario - revert. Otherwise, if everything is as expected, this code will be removed.